### ISSUE TITLE
JVM Segfault in libz3

### ISSUE THREAD

**Issue by bmmoore** [kframework/k-legacy#1634]
JVM crashes like this seem to be getting more common. It happened three times in a row trying to build a recent pull request

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00002b4cd3730106, pid=810, tid=47609081427712
#
# JRE version: Java(TM) SE Runtime Environment (8.0_45-b14) (build 1.8.0_45-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.45-b02 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libz3.so+0xf09106]  scoped_timer::imp::sig_handler(sigval)+0x6
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
```

Full logs at

http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/1727/artifact/k-distribution/hs_err_pid810.log

http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/1726/artifact/k-distribution/hs_err_pid20930.log

http://office.runtimeverification.com:8080/rv-jenkins/job/k-framework-pull-requests/1726/artifact/k-distribution/hs_err_pid20930.log
(this one doesn't recognize the stack frame, but register values look like we're in Z3)

@andreistefanescu, IIRC you built our libz3. Can you figure out what's causing this?

**Comment by andreistefanescu** [comment]
This is a known bug in libz3, caused by a race between the timeout handle and the some cleanup code. You can use `--z3-executable` which creates a new process instead of the jni.

**Comment by bmmoore** [comment]
I can, but that doesn't help in Jenkins. Looks like #1633 has failed on this four times in a row now.

**Comment by bmmoore** [comment]
Please link to the upstream issue.

**Comment by bmmoore** [comment]
@dwightguth is there any way to work around this in Jenkins?

**Comment by dwightguth** [comment]
@andreistefanescu this is causing too much disruption, it's time to fix the underlying cause. We need to either get the z3 guys to fix the segfault, or else remove all calls to the library from ktest and make the executable the default.

**Comment by andreistefanescu** [comment]
@dwightguth I disagree. The problem is that ktest and the tests are in the same process. Yes, ideally, the z3 guys would fix it, but we have very little control over it. I will keep the library as the default option, but if you want you can use `-z3-executable` for the remaining tests.

**Comment by bmmoore** [comment]
@andreistefanescu I don't see how that would stop this bug from leading to Jenkins build failures.

**Comment by edgar-pek** [comment]
@andreistefanescu https://github.com/andreistefanescu Just out of
curiosity (and because of strange nostalgia for decision procedure hacking)
how complex is the z3 issue?

On Tue, Aug 18, 2015 at 4:38 PM, Andrei Stefanescu <notifications@github.com

**Comment by grosu** [comment]
Guys, does that mean that if somebody clones the K sources from github in order to play with K and they type `mvn package`, then they will get errors due to z3?  If that is the case, then that is VERY BAD for us!  Most of the tests have nothing to do with z3.  And those having something to do with z3 should be made to work, or if they crash due to z3 errors, then we should report an error message saying so (otherwise people will think that K is the buggy one), but even that is still bad for K.

@andreistefanescu not sure I understand, does the --z3-executable solves this problem?  If yes, then what is the drawback of making it the default?  Do we pay any performance cost?

Let me use this track to tell you that I had a good meeting with Cesare Tinelli recently and he strongly encouraged us to use CVC4 as an SMT solver in K, because it does better than Z3 in many instances.  Maybe it is risky to switch completely, but we should at least allow it as an option to users.  Does this have anything to do with our current design of how we invoke z3?

**Comment by bmmoore** [comment]
So far it only seems to affect `mvn verify`.

**Comment by andreistefanescu** [comment]
@grosu this issue only affects `mvn verify`, not `mvn package`. The workaround is to run z3 in a separate process instead of using the z3 library, which is what `--z3-executable`. The downside is the crating new processes is expensive.

We can use cvc4 as an alternative smt solver if has java bindings and supports e-matching triggers.

@edgar-pek we set a timeout for z3 queries that check if path conditions are satisfiable. There is a race in z3 between the timeout implementation and some cleanup code, which results in segfaults sometimes.
